### PR TITLE
mail: read and act on Gmail in conversations (PR2)

### PR DESCRIPTION
**Pull request by srid**
## What changes

Implements **PR2 of #606**: agents can list and search Gmail, read full threads, save attachments, archive, trash, restore, apply existing labels and mark threads read. Each tool has a conversation story. Gmail remains the mail record; these tools do not write to the vault.

The pinned Himalaya fetched full MIME payloads but discarded them from JSON. A small Nix patch retains the typed payload; `mail-surface` now compares generated schemas with the checked-in decoder contracts. Claude and omp also needed to preserve the MCP server's refusal sentences when their adapters discard `structuredContent`.

### Ownership

All live capabilities remain inside mail's activation. Direct tool closures enter the Effect bridge's activation gate. Withdrawal cuts and joins calls; the runner kills and joins interrupted child processes before attachment and config cleanup. Four permits limit tool spawns; per-thread permits protect the complete before/write/after sequence. Account refreshes do not use the tools' spawn permits. `needs`, `wire.ts` and `faces` are unchanged; `Vault` is not required.

## Review fixes and refactor

- Reads refresh unknown label IDs once and preserve unresolved IDs; requested unknown label names still refuse writes.
- Claude and omp share `@olai/surface.refusalIn`. The real MCP server test suite compares parsed text with structured refusal details, so formatter changes fail the contract test.
- Missing attachments name the attachment, with command-specific translation at the Himalaya invocation boundary.
- Verb descriptions are readable clauses; the payload patch records its upstream issue and removal condition; attachment sizes round to one decimal KiB.

Separate passes using the [Hickey–Löwy approach](https://kolu.dev/blog/hickey-lowy/) identified duplicate account/refusal logic and a tool module mixing registration policy with mailbox implementation. `mailbox.ts` now owns commands, label caching and serialized mutations; `tools.ts` owns schemas, invocation gating and reply decoration. The account guard has one definition. SDK envelope extraction stays in each engine, while MCP text compatibility has one shared home and a test against its actual formatter. These are internal and pure-code boundaries: no new live service or owner, and the activation's cleanup ordering is preserved.

## Validation

**`just ci` passed** at [`d03cc1c88`](https://github.com/juspay/olai/commit/d03cc1c881b622e17d33d677d7c4307ac014a9a2). Odu run `0mu3f0u24-g2pmhgqq`: **67 passed, 0 failed, 0 skipped, 0 cancelled**, including `mail-surface`, the Nix build, all unit/typecheck shards and **1,825 browser scenarios**.

A real serve's `tools/list`, filtered to mail, returned exactly:

```text
mail_archive
mail_attachment
mail_inbox
mail_label
mail_read
mail_search
mail_thread
mail_trash
mail_untrash
```

`packages/server/src/profiles.test.ts` also verifies withdrawal and restoration over two off/on cycles. Unit coverage includes exact 64 KiB truncation, UTF-8 boundaries, the 50 MB preflight limit, four-child concurrency, serial writes, absent/faulted/expired accounts, unknown labels, actual child-process interruption, retained closures after shutdown, and schema validation of fake replies.

The existing agenda deletion assertion now waits for the server revision before checking absence (`packages/plugins/journal/e2e/steps/agenda_steps.ts`), fixing a race exposed by full CI.

### Browser scenarios

| Scenario | Result |
| --- | --- |
| Inbox lists three threads with unread state | Passed |
| Search finds the unread thread | Passed |
| HTML-only thread preserves its body and sender | Passed |
| Attachment is temporary and removed when mail stops | Passed |
| Unknown thread names the connected account | Passed |
| Archive removes a thread from inbox | Passed |
| Labels resolve names and reject an unknown label | Passed |
| Trash and restore change trash search results | Passed |
| Read marks flip unread both ways | Passed |
| Every tool refuses without an account and spawns nothing | Passed |
| A revoked grant refuses tools with the account reason | Passed |
| Deleted and unlisted label ids remain readable | Passed |
| A vanished attachment names the attachment in its refusal | Passed |

## PR2 checklist and implementation files

Paths beginning `src/` below are under `packages/plugins/mail/`.

### Tools and shared rules

- [x] `mail_inbox`: unread filtering, 1–50 bounds, default 20, metadata rows and pagination — `src/tools.ts`, `src/mailbox.ts`, `src/himalaya/threads.ts`.
- [x] `mail_search`: untouched Gmail query, pagination and spam/Trash flag — `src/tools.ts`, `src/mailbox.ts`.
- [x] `mail_thread`: message headers, labels, plain text/raw HTML, MIME traversal, 64 KiB limits and attachment metadata — `src/himalaya/threads.ts`, `src/tools.ts`, `src/mailbox.ts`.
- [x] `mail_attachment`: private runtime directory, safe capped filename, size preflight, returned path/size/MIME and scope cleanup — `src/attachments.ts`, `src/attachments.test.ts`.
- [x] `mail_archive`: remove INBOX, reread and report actual delta — `src/tools.ts`, `src/mailbox.ts`.
- [x] `mail_trash` and `mail_untrash`: thread-level verbs and reread — `src/tools.ts`, `src/mailbox.ts`.
- [x] `mail_label`: existing names, 1–10 per list, at least one list, resolve all names before mutation — `src/tools.ts`, `src/mailbox.ts`, `src/labels.ts`.
- [x] `mail_read`: add/remove UNREAD and report actual delta — `src/tools.ts`, `src/mailbox.ts`.
- [x] Label resolution: activation cache, both directions, case-insensitive names, refresh once on a miss — `src/labels.ts`, `src/labels.test.ts`.
- [x] Refusals: usable-token checks at entry and spawn, account reasons, unknown-thread wording, JSON error precedence and ops failure mapping — `src/tools.ts`, `src/mailbox.ts`, `src/account.ts`, `src/himalaya/run.ts`. Attachment preflight and remote-outcome limits are noted below.
- [x] Concurrency: four spawn permits, per-thread transaction permits, activation withdrawal — `src/tools.ts`, `src/mailbox.ts`, `src/tools.test.ts`, `packages/effect-cordis/src/index.ts`.
- [x] Logging: debug verb/args/duration/outcome and info write deltas, without bodies — `src/tools.ts`, `src/mailbox.ts`.

### Wire, package and browser

- [x] Seven additional frozen verbs and `GMAIL_VERBS` — `src/himalaya/verbs.ts`.
- [x] Pure row/full-thread/delta decoders with checked-in schemas — `src/himalaya/threads.ts`, `src/himalaya/schemas/`, `src/himalaya/threads.test.ts`.
- [x] New labels/tools/attachments modules — `src/labels.ts`, `src/tools.ts`, `src/attachments.ts`.
- [x] Register tools on the existing surface; activation-owned acquisition and cleanup; no new required service — `src/server.ts`.
- [x] Account cell and procedure faces remain unchanged — `src/wire.ts` (unchanged).
- [x] `tool.reply` with `fileOf: () => null`, all nine stories, refusal stories and `data-mail-story` — `src/browser.tsx`, `src/browser/story.ts`, `src/browser/story.tsx`, `src/testids.ts`.
- [x] No new page, inbox UI, pill or panel redesign — browser additions are the tool story only.

### Tests

- [x] Fake Himalaya: fixture mailbox, labels, paging/query support, mutations, file downloads and invocation recording — `src/appliance/testlib/fake-himalaya.ts`, `src/appliance/testlib/fixtures.ts`. Replies validate against the generated schemas in `src/appliance/testlib/testlib.test.ts`.
- [x] Fake ACP dispatches all required mail phrases through `useTool` and echoes replies — `packages/tests/agent/fake-acp-agent.ts`.
- [x] Decoder, labels, tool/refusal/concurrency, attachment and story unit tests — `src/himalaya/threads.test.ts`, `src/labels.test.ts`, `src/tools.test.ts`, `src/attachments.test.ts`, `src/browser/story.test.ts`; real cancellation in `src/himalaya/run.test.ts`.
- [x] All eleven required browser workflows plus stale-label and vanished-attachment edge cases — `e2e/features/mail_tools.feature`, `e2e/steps/mail_tools_steps.ts`, `packages/tests/fixtures/mail/mail.olai`.
- [x] Nine MCP tools appear and withdraw with the row — `packages/server/src/profiles.test.ts`.
- [x] `just mail-surface` checks verbs and exported schemas — `scripts/check-himalaya-surface.ts`.

### Docs

- [x] Current opening and “Asking an agent about your mail,” examples, label names, archive/Trash, permanent-delete exclusion, attachment lifetime and refused writes — `packages/plugins/mail/docs.md`.
- [x] Mail under **olai ✓** beside kolu/odu — `docs/chat.md`.
- [x] Reading and filing mail from a conversation — `docs/index.md`.

## Not done / specification corrections

- The pinned CLI uses `json-schema --dir <directory>`, and only structured readers publish schemas. Mutation and attachment-save commands return JSON strings, so no nonexistent schemas were invented for them. The four available schemas are checked against the built binary. The MIME payload fix is an Olai Nix patch tracked by [pimalaya/himalaya#750](https://github.com/pimalaya/himalaya/issues/750); remove it once the pin includes that fix.
- Fake writes persist in per-thread files inside the fake's disposable directory. Each invocation is a separate process; process-local memory cannot survive a subsequent `get`. Per-thread files also avoid overwriting an unrelated concurrent write.
- Attachments must first be discovered with `mail_thread`, because the specified attachment arguments do not carry the size/MIME and the frozen verbs have no message lookup. An unlisted attachment receives a “read the thread first” refusal without a fetch; a Gmail 404 after discovery says “this attachment is not on that message.”
- A remote mutation and its follow-up read cannot be one atomic transaction. Preflight refusals change nothing. If the mutation succeeds but the reread fails, the reply explicitly says the write was sent and asks for verification before retrying; the docs do not promise rollback.
- As specified, live thread properties/chips, doorbells, an inbox page, HTML conversion, permanent delete, sending and drafts remain outside this PR.

### CODE REVIEW

File: packages/plugins/mail/src/labels.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    if (names.some(n => !find(n))) yield* refresh
+    const missing = names.find(n => !find(n))
+    if (missing !== undefined) return yield* Effect.fail(new MailRefusal({ reason: `this mailbox has no label ${JSON.stringify(missing)}; it has: ${labels?.map(l => l.name).join(", ")}` }))
+    return names.map(n => find(n)!.id)
+  }))
+  const ensureIds = (ids: ReadonlyArray<string>) => permit.withPermits(1)(Effect.gen(function*() {
+    if (!labels) yield* refresh
+    const missing = () => ids.find(id => !labels?.some(label => label.id === id))
+    if (missing() !== undefined) yield* refresh
+    if (missing() !== undefined) return yield* Effect.fail(new MailRefusal({ reason: `this mailbox has no name for label id ${missing()}` }))
```
Comment (srid) [inline packages/plugins/mail/src/labels.ts]: **A read should not fail over a nameless id.** `get()` calls `ensureIds` on every thread it decodes, and this arm refuses the whole call when an id is still unknown after one refresh. Two ordinary ways to get here: a label deleted in Gmail between a listing and its per-thread `get` (the refresh happens, the id is gone, the thread still carries it for a while), and any system id `labels list` does not enumerate. Either turns `mail_inbox` into a refusal for the whole page.

`names()` already answers the raw id when there is no name, which is the right behaviour for a read. Suggest `ensureIds` refreshes once on a miss and then returns, never fails; keep the failure for `resolve`, where a name the person typed really is an error.

File: packages/plugins/claude/src/leg.ts [hunk 2]
```diff
@@ -809,5 +809,10 @@ export function replyIn(rawOutput: unknown): Record<string, unknown> | undefined
     : Array.isArray(rawOutput) ? rawOutput.find(block => record(block)?.["type"] === "text")?.text
     : undefined
   if (typeof text !== "string") return undefined
-  try { return record(JSON.parse(text)) } catch { return undefined }
+  try { return record(JSON.parse(text)) } catch {
+    // Claude forwards only MCP text blocks, dropping structuredContent. The
+    // owned MCP server's branded refusal still carries its kind and reason.
+    const refusal = /^surface-mcp: `[^`]+` was refused \((usage|not-found|validation|busy)\): ([\s\S]+)$/.exec(text)
```
Comment (srid) [inline packages/plugins/claude/src/leg.ts]: **One regex, two legs, and a pin's wording.** This and `omp/src/leg.ts:222` are the same expression, and both are a literal transcription of how `@kolu/surface-mcp` spells a refusal today. If a kolu pin bump rewords it (adds a kind, changes the backticks, drops the prefix), both adapters go back to `undefined` and every mail refusal story disappears with nothing red anywhere: the legs' tests match the literal, not the formatter.

Suggest one `refusalIn(text)` in a package both legs already import from (chat's contract, or `@olai/surface` beside the MCP glue), with a test that asks the real `surface-mcp` refusal formatter for a sentence and parses it back. Then the day the wording moves, one test fails in the right place.

File: packages/plugins/mail/src/tools.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    }),
+    entry => entry.permit.withPermits(1)(work),
+    entry => Effect.sync(() => { if (--entry.users === 0) writes.delete(id) }),
+  )
+  const perform = <A>(verb: string, args: unknown, work: Effect.Effect<A, MailRefusal>) => Effect.gen(function*() {
+    const started = Date.now()
+    const result = yield* owned.through(Effect.result(work), fiber => fiber ? Fiber.join(fiber) : Effect.succeed(Result.fail(new MailRefusal({ reason: "mail is no longer running" }))))
+    yield* Effect.logDebug(`mail tool verb=${verb} args=${JSON.stringify(args)} duration=${Date.now() - started} outcome=${result._tag === "Success" ? "ok" : "refused"}`)
+    if (result._tag === "Failure") {
+      const reason = !result.failure.reason.startsWith("the write was sent") && /\b404\b|not found|notFound/i.test(result.failure.reason) ? `this thread is not in ${machine.current().address}` : result.failure.reason
```
Comment (srid) [inline packages/plugins/mail/src/tools.ts]: Nit: the 404 heuristic rewrites every not-found into *this thread is not in <address>*, including an `attachments get` miss (the fake's own `404 not found` for an unknown attachment lands here). Pass the subject in from the tool, so an attachment miss says *this attachment is not on that message*.

File: packages/plugins/mail/src/himalaya/verbs.ts [hunk 4]
```diff
@@ -73,6 +73,13 @@ export const GMAIL = {
     path: ["gmail", "profile", "get"],
     says: "the signed-in address and the mailbox's totals",
   },
+  threadsList: { id: "threads.list", path: ["gmail", "threads", "list"], says: "threads.list" },
```
Comment (srid) [inline packages/plugins/mail/src/himalaya/verbs.ts]: Nit: `says` is meant to be the sentence a reader gets, and it is what the fake's `ANNOUNCES` and the surface check's failure print. With the id in it they read *`himalaya gmail threads list` (threads.list, threads.list)*. One clause each, like `profileGet` has.

File: nix/himalaya.nix [hunk 5]
```diff
@@ -53,5 +53,9 @@ in
       inherit pkgs;
       system = pkgs.stdenv.hostPlatform.system;
     };
-  };
+  }).overrideAttrs (old: {
+    # The pin fetches full MIME payloads but drops them while rendering JSON.
+    # Retain that typed field; mail-surface checks the generated schema.
+    patches = (old.patches or [ ]) ++ [ ./himalaya-thread-payload.patch ];
```
Comment (srid) [inline nix/himalaya.nix:60]: Please give the patch provenance: the upstream issue or PR at pimalaya/himalaya that carries this change, in the comment here and at the head of `himalaya-thread-payload.patch`. The schema check will fail the day the pin includes it, and whoever bumps the pin then needs to know that dropping the patch is the fix rather than a regression. If no upstream PR exists yet, opening one is worth the ten minutes: this is a one-line `payload: message.payload` that upstream almost certainly wants.

File: packages/plugins/mail/src/browser/story.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    return { kind, text: `${r.threads.length} threads ${kind === "inbox" ? "in INBOX" : `for ${JSON.stringify(r.query)}`}` }
+  }
+  if (kind === "thread") {
+    if (typeof r.subject !== "string" || !Array.isArray(r.messages)) return null
+    const from = record(r.messages[0])?.from
+    return { kind, text: `${r.subject} · ${typeof from === "string" ? from : ""} · ${r.messages.length} messages` }
+  }
+  if (kind === "attachment") {
+    if (typeof r.filename !== "string" || typeof r.bytes !== "number") return null
+    return { kind, text: `${r.filename} · ${r.bytes >= 1024 ? `${r.bytes / 1024} KiB` : `${r.bytes} bytes`}` }
```
Comment (srid) [inline packages/plugins/mail/src/browser/story.ts]: Nit: `r.bytes / 1024` prints *12.6953125 KiB* for anything that is not a multiple of 1024. Round to one decimal, or reuse whatever the attachments strip in chat already uses for sizes.

**Review by srid — COMMENTED** [review]
Reviewed against the PR 2 spec in #606 (the folded read-and-write version) and the Cordis rules.

**Verdict: this is right, and it is complete.** All nine tools are there with the shapes the spec asked for; every write re-reads and reports the real delta; labels resolve by name and refuse before any write; attachments live in a mail-owned directory that dies with the row; the fake mailbox mutates and validates against Himalaya's own schemas; the eleven scenarios are all present and named in the body; the profiles test proves advertise and withdraw over two cycles. The Cordis story is stronger than the spec asked for: tool closures go through the activation `gate`, the runner kills and joins its child on interrupt, and the finalizer order (fibers, attachments, config) is stated and tested. The five documented deviations are all correct calls, and the Himalaya payload patch is the honest fix for a binary that fetched bodies and threw them away.

Six things below, none blocking. The first two are worth doing before merge; the rest are nits that can ride the next PR.

1. A **read fails when a label id has no name** (`labels.ensureIds`): a label deleted between two calls, or any id `labels list` does not carry, makes `mail_inbox`, `mail_search` and `mail_thread` refuse outright. `names()` already falls back to the raw id, so the read should too.
2. The **refusal parser in the Claude and omp legs** is one regex duplicated twice and coupled to `@kolu/surface-mcp`'s wording. A pin bump that rewords the refusal silently turns every mail refusal story back into nothing. One shared parser, tested against the real formatter, would make the next bump fail loudly instead.
3. The 404 heuristic names *this thread* for an attachment miss too.
4. `verbs.ts` fills `says` with the id, so the fake's announcement and the surface check's failure read *threads.list (threads.list, threads.list)*.
5. The Himalaya patch has no provenance; link the upstream issue or PR in `himalaya.nix` so a future pin bump knows when to drop it.
6. The attachment story prints unrounded KiB.

The `agenda_steps.ts` wait is the right fix for the flake PR 1 named. Good to merge once the odu legs on this head report (the body says `just ci` passed at this commit; the GitHub rollup had not caught up when I looked).

**Comment by srid** [comment]
Re-checked at `d03cc1c88`. Every point from the review is closed in code, with a test each:

- reads survive a nameless label id (`ensureIds` refreshes once and returns; `labels.test.ts` and a new `@mail-himalaya:stale` scenario prove it, and `resolve` still refuses a typed name);
- the refusal parser is one `refusalIn` in `@olai/surface`, both legs use it, and `packages/server/src/mcp/tools.test.ts` round-trips the real formatter's output for every refusal kind the suite exercises, so a kolu pin bump that rewords it fails there;
- a 404 is translated where the verb is known, so an attachment miss names the attachment (new scenario);
- `says` reads as sentences; the patch and `himalaya.nix` cite pimalaya/himalaya#750; KiB is rounded.

The `mailbox.ts` / `tools.ts` split is a good boundary and the cordis.md note describes it accurately.

One nit for whenever `docs.md` is next touched: the last sentence of the new paragraph (*Attachment stories round KiB sizes to one decimal place*) is an implementation detail on the user page; the rest of that paragraph belongs there.

Nothing further from me. Merge once the odu legs report on this head.
